pydantic/dataclasses.py

```diff
@@ -33,136 +76,210 @@ def __get_validators__(cls: Type['Dataclass']) -> 'CallableGenerator':
+    _cls: Optional[Type[Any]] = None,
+    *,
+    init: bool = True,
+    repr: bool = True,
+    eq: bool = True,
+    order: bool = False,
+    unsafe_hash: bool = False,
+    frozen: bool = False,
+    config: Union[ConfigDict, Type[Any], None] = None,
+    validate_on_init: Optional[bool] = None,
+) -> Union[Callable[[Type[Any]], 'DataclassClassOrWrapper'], 'DataclassClassOrWrapper']:
+    """
+    Like the python standard lib dataclasses but with type validation.
+    The result is either pydantic dataclass that will validate input data
+    or a wrapper that will trigger validation around a stdlib dataclass
+    to avoid modifying it directly
+    """
+    the_config = get_config(config)
+
+    def wrap(cls: Type[Any]) -> 'DataclassClassOrWrapper':
         import dataclasses
 
-        return cls(**dataclasses.asdict(v))
-    else:
-        raise DataclassTypeError(class_name=cls.__name__)
+        if is_builtin_dataclass(cls):
+            should_validate_on_init = False if validate_on_init is None else validate_on_init
+            _add_pydantic_validation_attributes(cls, the_config, should_validate_on_init, '')
+            return DataclassProxy(cls)
 
+        else:
+            dc_cls_doc = cls.__doc__ or ''  # needs to be done before generating dataclass
+            dc_cls = dataclasses.dataclass(  # type: ignore
+                cls, init=init, repr=repr, eq=eq, order=order, unsafe_hash=unsafe_hash, frozen=frozen
+            )
+            should_validate_on_init = True if validate_on_init is None else validate_on_init
+            _add_pydantic_validation_attributes(dc_cls, the_config, should_validate_on_init, dc_cls_doc)
+            return dc_cls
 
-def _get_validators(cls: Type['Dataclass']) -> 'CallableGenerator':
-    yield cls.__validate__
+    if _cls is None:
+        return wrap
 
+    return wrap(_cls)
 
-def setattr_validate_assignment(self: 'Dataclass', name: str, value: Any) -> None:
-    if self.__initialised__:
-        d = dict(self.__dict__)
-        d.pop(name, None)
-        known_field = self.__pydantic_model__.__fields__.get(name, None)
-        if known_field:
-            value, error_ = known_field.validate(value, d, loc=name, cls=self.__class__)
-            if error_:
-                raise ValidationError([error_], self.__class__)
 
-    object.__setattr__(self, name, value)
+@contextmanager
+def set_validation(cls: Type['DataclassT'], value: bool) -> Generator[Type['DataclassT'], None, None]:
+    original_run_validation = cls.__pydantic_run_validation__
+    try:
+        cls.__pydantic_run_validation__ = value
+        yield cls
+    finally:
+        cls.__pydantic_run_validation__ = original_run_validation
 
 
-def is_builtin_dataclass(_cls: Type[Any]) -> bool:
+class DataclassProxy:
+    __slots__ = '__dataclass__'
+
+    def __init__(self, dc_cls: Type['Dataclass']) -> None:
+        object.__setattr__(self, '__dataclass__', dc_cls)
+
+    def __call__(self, *args: Any, **kwargs: Any) -> Any:
```

Comment:
I guess we lose the intellisense/ide autocompletion for the underlying class init parameters in addition to static type checking with this approach, don't we?

Reply:
Also, do we lose `isinstance(wrapper_instance,wrapper_instance.__dataclass__.__class__)` with this proxy?

Reply:
Yep currently we lose intellisense. I tried to have a `if TYPE_CHECKING` block with a generic but didn't see anything in typing to mimic the `__call__` to the  one of `T`.
If you have any clue feel free to add it!

For the second question this is what we have
```py
from dataclasses import dataclass as sdlib_dataclass

from pydantic.dataclasses import dataclass as py_dataclass


@sdlib_dataclass
class A:
    x: int

a = A(x=1)

A_PY = py_dataclass(A)
validated_a = A_PY(x='1')

assert validated_a.x == 1
assert isinstance(validated_a, A)
assert not isinstance(validated_a, A_PY)
```

Reply:
mypy doesn't recognising a pydantic dataclass as dataclass with this, but from my tests it doesn't on master/v1.8 either.

It would be great if there was a way to get this to work out of the box. But since it's not a regression, it shouldn't block this PR.